tardis/simulation/base.py

```diff
@@ -410,3 +414,161 @@ def from_config(cls, config, **kwargs):
 
+    def make_source_function(self):
+        """
+        Calculates the source function using the line absorption rate estimator `Edotlu_estimator`
+
+        Formally it calculates the expression ( 1 - exp(-tau_ul) ) S_ul but this product is what we need later,
+        so there is no need to factor out the source function explicitly.
+
+        Parameters
+        ----------
+        model : tardis.model.Radial1DModel
+
+        Returns
+        -------
+        Numpy array containing ( 1 - exp(-tau_ul) ) S_ul ordered by wavelength of the transition u -> l
+        """
+        model = self.model
+        plasma = self.plasma
+        runner = self.runner
+        atomic_data = self.plasma.atomic_data
+
+        Edotlu_norm_factor = (1 / (runner.time_of_simulation * model.volume))
+        exptau = 1 - np.exp(- plasma.tau_sobolevs)
+        Edotlu = Edotlu_norm_factor * exptau * runner.Edotlu_estimator
+
+        upper_level_index = atomic_data.lines.set_index(['atomic_number', 'ion_number', 'level_number_upper']).index.copy()
+        e_dot_lu          = pd.DataFrame(Edotlu, index=upper_level_index)
+        e_dot_u           = e_dot_lu.groupby(level=[0, 1, 2]).sum()
+        e_dot_u.index.names = ['atomic_number', 'ion_number', 'source_level_number'] # To make the q_ul e_dot_u product work, could be cleaner
+        transitions       = atomic_data.macro_atom_data[atomic_data.macro_atom_data.transition_type == -1].copy()
+        transitions_index = transitions.set_index(['atomic_number', 'ion_number', 'source_level_number']).index.copy()
+        tmp  = plasma.transition_probabilities[(atomic_data.macro_atom_data.transition_type == -1).values]
+        q_ul = tmp.set_index(transitions_index)
+        t    = model.time_explosion.value
+        wave = atomic_data.lines.wavelength_cm[transitions.transition_line_id].values.reshape(-1,1)
+        att_S_ul =  ( wave * (q_ul * e_dot_u) * t  / (4*np.pi) )
+
+        result = pd.DataFrame(att_S_ul.as_matrix(), index=transitions.transition_line_id.values)
+        return result.ix[atomic_data.lines.index.values].as_matrix(),e_dot_u
+
+    def integrate(self):
+        model = self.model
+        plasma = self.plasma
+        runner = self.runner
+
+        num_shell, = self.runner.volume.shape
+        ps         = np.linspace(0.999, 0, num = 20) # 3 * num_shell)
+        R_max      = runner.r_outer_cgs.max()
+        R_min_rel  = runner.r_inner_cgs.min() / R_max
+        ct         = c.c.cgs.value * model.time_explosion.value / R_max
+
+        att_S_ul, Edot_u = self.make_source_function()
+        J_blues    = plasma.j_blues
+        J_rlues    = plasma.j_blues * np.exp( -plasma.tau_sobolevs.values) + att_S_ul
+
+        r_shells = np.zeros((num_shell+1,1))
+        # Note the reorder from outer to inner
+        r_shells[1:,0],r_shells[0,0] = self.runner.r_inner_cgs[::-1] / R_max, 1.0
+        z_crossings = np.sqrt(r_shells**2 - ps**2)
+
+        z_ct = z_crossings/ct
```

Comment:
I can fix all PEP8 violations introduced by this PR or I could work on the proper version which removes 80% of the code introduced by this PR anyway.

I'm perfectly fine fixing PEP8 but I don't think we gain much for the time this takes.